### ISSUE TITLE
Move execution out of Container into an Executor Protocol

### ISSUE THREAD

**Issue by adriangb** [adriangb/di#12]
As a possible path for #11, we could move execution out of the container.

We could have an `Executor` interface that takes a DAG of `Tasks and executes them.
Then `di` can provide several implementations of an `Executor`, for example:
- Move sync deps into a thread, parallelize -> good for IO bound workflows w/ sync deps that do a lot of IO
- Don't move sync deps into a thread, parallelize -> good for IO bound workflows w/ sync deps that do not do much
- Only sync, run in threads -> a sync interface, good for IO bound workflows
- Only sync, run without threads -> a sync interface, good for simple cases where there is no IO

`Executor` could be a constructor parameter to `Container` so that users can do something like:

```python
from di import Container, SingleThreadedExecutor
from webframework import App

def get_app() -> App:
    container = Container(executor=SingleThreadedExecutor())
    return App(container=container)
```

**Comment by adriangb** [comment]
Well here we are, I was able to come up with an (admittedly not great) solution in c31b3a22201250f110c344ca89ca20957fbacb11.

I think the overall approach is pretty solid, but the interfaces / composition needs to be cleaned up a bit.

The TLDR for now is that there is `Container.execute_sync` and `Container.execute_async`, and `Container.__init__` takes an executor parameter, which correspond to an `Executor` interface that takes the DAG of tasks and executes it using whatever strategy it wants, and it can support async dependencies or not.

This resolves #11 since it is now possible to have a fully sync single threaded interface

**Comment by adriangb** [comment]
This is now completed. The [interface](https://github.com/adriangb/di/blob/main/di/types/executor.py) and [default implementation](https://github.com/adriangb/di/blob/main/di/executors.py) are pretty clear.
